### DISCUSSION TITLE
How to generate openapi spec with Path, Query and Header Parameters as Components to avoid repetition

### DISCUSSION THREAD

**Discussion by SPFZ** [fastapi/fastapi#13049] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
import uuid as UUID
from fastapi import FastAPI, Path, Depends
from pydantic import BaseModel, Field

app = FastAPI()


class ItemId(BaseModel):
    id: UUID.UUID = Field(
        Path(
            description="uuid to identify items"
        )
    )


@app.get("/items/{item_id}")
async def read_item(item_id: ItemId = Depends()):
    """GET an Item by uuid"""
    return {"item_id": item_id}


@app.delete("/items/{item_id}")
async def delete_item(item_id: ItemId = Depends()):
    """DELETE an Item by uuid"""
    return {"item_id": item_id}
```

### Description

tl;dr: How to get Query-/Path-/Header-parametes to show up as **components** in the openapi spec?

I am developing an API based on an openapi spec that I got from the project. I am trying to make the diff between the spec I got and the spec that fastapi generats as small as possible - ideally it should be identical.

I have a set of default Headers (e.g. Cache-Control) and Path parameters (uuid) that are used on many enpoints. Because they are reused often the original spec uses components to specify them and then just uses `$refs`. 

The example code generates the item_id parameter for each endpoint:
```json
{
    [...]
    "paths": {
        "/items/{item_id}": {
            "get": {
                "summary": "Read Item",
                "description": "GET an Item by uuid",
                "operationId": "read_item_items__item_id__get",
                "parameters": [
                    {
                        "name": "id",
                        "in": "path",
                        "required": true,
                        "schema": {
                            "type": "string",
                            "format": "uuid",
                            "description": "uuid to identify items",
                            "title": "Id"
                        },
                        "description": "uuid to identify items"
                    }
                ],
[...]
}
```

I would like it to be separated as a component. (This works for Requests and Responses but not for Path, Query and Header parametes.)

I should look something like this:
```json
{
    [...]
    "paths": {
        "/items/{item_id}": {
            "get": {
                "summary": "Read Item",
                "description": "GET an Item by uuid",
                "operationId": "read_item_items__item_id__get",
                "parameters": [
                    - $ref: "#/components/parameters/UuidInPathParam"
                ],
[...]
components:
  headers: [...]
  parameters:
    UuidInPathParam:
      description: A universally unique identifier (UUID) according to RFC 4122.
      in: path
      name: id
      required: true
      schema:
        title: Id
        type: string
        format: uuid
        description: A universally unique identifier (UUID) according to RFC 4122.
}
```

I could not find a way to do this in a way that works. I known that I could customize to openapi gerneration: [extending-openapi](https://fastapi.tiangolo.com/how-to/extending-openapi)

### Operating System

Linux

### FastAPI Version

0.115.6

### Pydantic Version

2.10.3

### Python Version

Python 3.12.7

**Comment by yvallois** [comment]
I don't think this is possible with the current version of FastApi.
The body_field attribute of each route is the only one that generates a definition that will finally generate a ref.
Of course you can probably use [extending-openapi](https://fastapi.tiangolo.com/how-to/extending-openapi/) but I don't think this is naturally managed by FastAPI.

**Comment by yakubka** [comment]
FastAPI generates path, query, and header parameters inline in each operation object and does not place them in `components/parameters` by default. The OpenAPI specification allows parameters to be defined in `components/parameters` and referenced with `$ref`, but FastAPI`s schema generator does not produce that structure.

If you need parameters as reusable components, you can post-process the generated spec.

```python
from fastapi.openapi.utils import get_openapi

def custom_openapi():
    if app.openapi_schema:
        return app.openapi_schema
    schema = get_openapi(title=app.title, version=app.version, routes=app.routes)

    # extract shared parameters into components/parameters
    schema.setdefault("components", {}).setdefault("parameters", {})
    for path_item in schema.get("paths", {}).values():
        for operation in path_item.values():
            if isinstance(operation, dict):
                for param in operation.get("parameters", []):
                    name = f"{param[name]}_{param[in]}"
                    schema["components"]["parameters"][name] = param

    app.openapi_schema = schema
    return schema

app.openapi = custom_openapi
```

This is a post-processing step and does not replace inline parameters with `$ref` entries, but it gives you the components section. Replacing inline params with refs requires more work and is not something FastAPI will do automatically.
